ports/esp32/machine_i2s.c

```diff
@@ -0,0 +1,440 @@
+        case ESP_ERR_INVALID_ARG:
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S driver install: Parameter error"));
+            break;
+        case ESP_ERR_NO_MEM:
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S driver install: Out of memory"));
+            break;
+        default:
+            // this error not documented in ESP-IDF
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S driver install: Undocumented error")); 
+            break;
+    }
+
+    i2s_pin_config_t pin_config;
+    pin_config.bck_io_num = self->bck;
+    pin_config.ws_io_num = self->ws;
+    pin_config.data_out_num = self->sdout;
+    pin_config.data_in_num = self->sdin;
+
+    ret = i2s_set_pin(self->id, &pin_config);
+    switch (ret) {
+        case ESP_OK:
+            break;
+        case ESP_ERR_INVALID_ARG:
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S set pin: Parameter error"));
+            break;
+        case ESP_FAIL:
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S set pin: IO error"));
+            break;
+        default:
+            // this error not documented in ESP-IDF
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S set pin: Undocumented error")); 
+            break;
+    }
+
+    self->used = true;
+}
+
+/******************************************************************************/
+// MicroPython bindings for I2S
+STATIC void machine_i2s_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    machine_i2s_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_printf(print, "I2S(id=%u, bck=%d, ws=%d, sdout=%d, sdin=%d\n"
+            "standard=%u, mode=%u,\n"
+            "dataformat=%u, channelformat=%u,\n"
+            "samplerate=%d,\n"
+            "dmacount=%d, dmalen=%d,\n"
+            "apllrate=%d)",
+            self->id, self->bck, self->ws, self->sdout, self->sdin,
+            self->standard, self->mode,
+            self->dataformat, self->channelformat,
+            self->samplerate,
+            self->dmacount, self->dmalen,
+            self->apllrate
+            );
+}
+
+STATIC mp_obj_t machine_i2s_make_new(const mp_obj_type_t *type, size_t n_pos_args, size_t n_kw_args, const mp_obj_t *args) {
+    mp_arg_check_num(n_pos_args, n_kw_args, 1, MP_OBJ_FUN_ARGS_MAX, true);
+
+    machine_i2s_obj_t *self;
+
+    // note: it is safe to assume that the arg pointer below references a positional argument because the arg check above
+    //       guarantees that at least one positional argument has been provided
+    i2s_port_t i2s_id = mp_obj_get_int(args[0]);
+    if (i2s_id == I2S_NUM_0) {
+        self = &machine_i2s_obj[0];
+    } else if (i2s_id == I2S_NUM_1) {
+        self = &machine_i2s_obj[1];
+    } else {
+        mp_raise_ValueError(MP_ERROR_TEXT("I2S ID is not valid"));
+    }
+
+    self->base.type = &machine_i2s_type;
+    self->id = i2s_id;
+
+    // is I2S peripheral already in use?
+    if (self->used) {
```

Comment:
Hi again Mike.

Is it a deliberate design choice to raise this exception in this way? I'm encountering the same error as reported by https://github.com/micropython/micropython/pull/4471#issuecomment-671231735 where I can't re-create my I2S object after a soft-reset because the port is reports as already in use (and in my use case, users can repeatedly upload new programs (similarly to the REPL + Ctrl-D) that triggers a soft reset of uPython).

I haven't yet thoroughly searched for some kind of static class method I can call to deinit/reset the I2S peripheral before I re-create my uPython I2S instance (and a quick scan of your examples repo seems to require an existing I2S object instance to call deinit() from, which I don't have post-reset) - so instead for now I just commented out this exception and now I can happily re-create the I2S object after a soft reset without issues. Doing this might however be a Bad Thing (tm) for reasons that are currently not obvious to me!

Feel free to let me know if I'm doing something stupid, but if not might I suggest a note under your esp32-i2s-examples about how to handle creating an I2S instance if the underlying hardware peripheral is still active from a previous initialization (ie: what should I do with the exception getting thrown if the I2S port is reported as already active?), and perhaps (maybe a horrible hack) a flag in the constructor parameters to choose to ignore the error and finish the constructor initialization code regardless?

Reply:
I can see how this exception is causing problems to upload new programs.  I eliminated all checks around mapping an I2S hardware peripheral to an I2S instance.  The exception no longer happens when re-uploading a program.  There is a downside to this change -- it's now possible to instantiate multiple I2S objects that use the same I2S hardware peripheral.  But, that situation doesn't cause problems, and other MicroPython machine modules already allow this. 

The PR has been updated.   I included a workaround for a bug that appears in a specific version of the ESP-IDF (v4.0.2).

Reply:
I would tie into soft-reset and reinit I2S, I believe this solves the "re-upload" issues. https://github.com/micropython/micropython/blob/master/ports/esp32/main.c#L115

Reply:
> I would tie into soft-reset and reinit I2S

That is totally the best way to solve the problem.  I didn't know there was a hook to do this.  Thanks a ton for reviewing this change and making the suggestion.

Reply:
The previous changes were backed out and a new module init routine was added, that is called from the point of soft reset.  A better implementation of a bug workaround was added as well.